p2p/module.go

```diff
@@ -31,6 +32,49 @@ type p2pModule struct {
 	network typesP2P.Network
 }
 
+// HandleEvent implements modules.P2PModule
+func (m *p2pModule) HandleEvent(message *anypb.Any) error {
+	switch message.MessageName() {
+	case messaging.BeforeHeightChangedEventType:
+		msg, err := codec.GetCodec().FromAny(message)
+		if err != nil {
+			return err
+		}
+		event, ok := msg.(*messaging.BeforeHeightChangedEvent)
+		if !ok {
+			return fmt.Errorf("failed to cast message to BeforeHeightChangedEvent")
+		}
+
+		m.broadcastAddressBookAtHeight(event.CurrentHeight)
+
+	default:
+		return typesP2P.ErrUnknownEventType(message.MessageName())
+	}
+	return nil
+}
+
+// broadcastAddressBookAtHeight broadcasts on the event bus the address book at the given height
+func (m *p2pModule) broadcastAddressBookAtHeight(height uint64) error {
```

Comment:
**Starting a discussion here.**

I'm thinking whether we actually need to broadcast this as an event or not.

The persistence layer (i.e. staked validators) are going to tell us what the validator set is. The address book is going to tell us how to reach them. What if `p2pModule` exposes a `GetAddressBook` function that is called synchronously. That way, consensus can compare the URL found through peer / churn when it needs to do a `send` during the HotPOKT lifecycle.

Reply:
I understand it now and correct me if I am wrong.

There's no need to broadcast the addressBook from P2P to other modules at every height. 
It would be necessary if we wanted to do something with it for some reason (I can think of observability but at the moment I can't think about anything else). Obviously it should not be part of the world state because, inherently, the addressBook is a "perspective" of the network from the current peer.

Reply:
> There's no need to broadcast the addressBook from P2P to other modules at every height.
It would be necessary if we wanted to do something with it for some reason

Yea, that's the only thing I'm imagining right now, but I'd prefer not to be broadcasting all this data until we are really using it.

If we were to use it for observability, what I would do is:

1. Create a function in the P2P interface called `GetAddressBook` that lets you retrieve it. It does not introduce new complexity, and can be multi-purposed. E.g. when we build a dashboard to view it. I bet I can asked ChatGPT to do this.

2. Create an event called "AddressBookChanged", that triggers us to poll.

> Obviously it should not be part of the world state because, inherently, the addressBook is a "perspective" of the network from the current peer.

Completely correct.